igcollect/sfp_dom.py

```diff
@@ -0,0 +1,372 @@
+    parser = ArgumentParser()
+    parser.add_argument('--prefix', default='sfp_dom')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    physical_interfaces = get_physical_interfaces()
+
+    for iface in physical_interfaces:
+        try:
+            data = ethtool_get_module_data(iface)
+            print_metrics(args.prefix, iface, data.metrics)
+        except (NoSFPException, UnsupportedSFPException) as exc:
+            print(exc, file=sys.stderr)
+            continue
+
+
+def get_physical_interfaces():
+    scn = '/sys/class/net'
+    physical_interfaces = []
+    for dev in listdir(scn):
+        dev_path = join(scn, dev)
+        if not (islink(dev_path) or isdir(dev_path)):
+            # Skip regular files
+            continue
+        # Physical interfaces have device as a symlink to /sys/class/device_id
+        if islink(join(scn, dev, 'device')):
+            physical_interfaces.append(dev)
+
+    return physical_interfaces
+
+
+def a2_offset_to_celsius(data, offset):
+    """
+    Returns a signed float built from a 16-bit signed twos complement value
+    in increments of 1/256 degrees Celsius
+    """
+    signed = ctypes.c_int16((data[SFF_A2_BASE + (offset)] << 8 | data[
+        SFF_A2_BASE + (offset) + 1]))
+    return signed.value / 256
+
+
+def a2_offset_to_volts(data, offset):
+    return a2_offset_to_int(data, offset) / 10000
+
+
+def a2_offset_to_milliamps(data, offset):
+    return a2_offset_to_int(data, offset) / 256
+
+
+def a2_offset_to_int(data, offset):
+    """
+    Returns an integer built from a 16-bit unsigned integer obtained from
+    data, at the specified offset
+    """
+    return (data[SFF_A2_BASE + (offset)] << 8 | data[
+        SFF_A2_BASE + (offset) + 1])
+
+
+def print_metrics(prefix, interface, metrics):
+    timestamp = int(time())
+    for metric, child in metrics.items():
+        for value_key, value in child.items():
+            print(
+                f'{prefix}.{interface}.{metric}.{value_key}'
+                f' {value} {timestamp}'
+            )
+
+
+def ethtool_get_module_data(interface):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    name = (ctypes.c_ubyte * IFNAMSIZ)(*bytearray(str(interface).encode()))
+    ifr = ifreq()
+    ifr.ifr_name = name
+
+    # Get module info
+    em = ethtool_modinfo()
+    em.cmd = ETHTOOL_GMODULEINFO
+    ifr.ifr_data.ethtool_modinfo_ptr = ctypes.pointer(em)
+
+    try:
+        fcntl.ioctl(sock, SIOCETHTOOL, ifr)
+    except OSError as exc:
+        if exc.errno == 95:
+            raise NoSFPException(f'Interface {interface} has no SFP module')
+        else:
+            raise
+
+    ee = ethtool_eeprom()
+    ee.cmd = ETHTOOL_GMODULEEEPROM
+    ee.len = em.eeprom_len
+    ee.offset = 0
+    ifr.ifr_data.ethtool_eeprom_ptr = ctypes.pointer(ee)
+
+    try:
+        fcntl.ioctl(sock, SIOCETHTOOL, ifr)
+    except OSError as exc:
+        raise
+
+    if em.type == ETH_MODULE_SFF_8079:
```

Comment:
you could refactor these elifs to something like `if em.type in (ETH_MODULE_SFF_8079, ...)`